### ISSUE TITLE
Available storage space not what was expected / calculated

### ISSUE THREAD

**Issue by Iguana9999phil** [minio/minio#11116]
I've got 8 nodes with 15 disks each, each disk got 9TB available. All disks are formatted in XFS. I've set Minio up with EC:2 for standard storage_class to get an average ratio of 1.33.

```storage_class standard=EC:2 rrs=EC:2 dma=write```

Minio created 15 sets of 8 disks on boot so with EC:2 I should have the following storage space available:

15 * 6 * 9TB = 810TB usable storage space on 1080TB raw. (raw=on disk usage)

After copying around 670TB of usable data, Minio is complaining that disks are full (Shouldn't there be a 900Mb threshold per disk, I had disks with MB's free) 

Reviewing the nodes show that most disks are around 98% full. Calculating the remaining raw free space over all nodes (on average) gives me 23TB free while I expected 180TB (raw on disk) to be free with EC:2 (around 1.33 ratio)

the .minio.sys dir per disk is around 10GB per disk as there are some tmp files and no multipart uploads are present.

Dataset was uploaded with rclone, in the first 100TB I've checked and done some calculations to make sure the ratio of 1.33 was getting hit, after that was confirmed I've let the copy proceed.

## Expected Behavior
Would expect 180TB of raw space to be available as of this moment.

## Current Behavior
Only 32TB of raw space is available as of this moment.

## Possible Solution
No idea, starting all over with the whole dataset wouldn't be preferred especially without knowing what the usable space will be.

## Steps to Reproduce (for bugs)

1. Set up minio 8 nodes, 15 disks per node, let minio create 15 sets of 8 disks.
2. Set standard storage_class to EC:2 using mc
3. Copy data
4. Check free space vs expected free space

## Context
I'm unable to copy all my data over due to missing a large amount of expected storage. Data set is 710TB total which should fit nicely in 810TB as I was expecting.

## Regression
No idea

## Your Environment

* Version used (`minio --version`):RELEASE.2020-10-27T04-03-55Z
* Server setup and configuration: 8 nodes, 15 disk per node.
* Operating System and version (`uname -a`): Debian 10 4.19.0-10-amd64 #1 SMP Debian 4.19.132-1 (2020-07-24) x86_64 GNU/Linux

**Comment by harshavardhana** [comment]
@Iguana9999phil can you provide `df -h` output from all nodes?

**Comment by Iguana9999phil** [comment]
See the attached file:
[nodes.txt](https://github.com/minio/minio/files/5699933/nodes.txt)

**Comment by harshavardhana** [comment]
Also `df -i` @Iguana9999phil  - also calcuate `du -sh <backend_disk>/.minio.sys/`  across all drives all all servers.

**Comment by Iguana9999phil** [comment]
See attached:
[inodes.txt](https://github.com/minio/minio/files/5700014/inodes.txt)

**Comment by harshavardhana** [comment]
> also calcuate `du -sh <backend_disk>/.minio.sys/` across all drives all all servers.

@Iguana9999phil and this one?

**Comment by harshavardhana** [comment]
@Iguana9999phil what is the maximum size of your object v/s smallest size of your object?

**Comment by Iguana9999phil** [comment]
> > also calcuate `du -sh <backend_disk>/.minio.sys/` across all drives all all servers.
> 
> @Iguana9999phil and this one?

It's there with the inodes.txt

No idea about the sizes of the objects, these are media files so ranging from 160GB to 600MB in sizes.

**Comment by harshavardhana** [comment]
NOTE: each object gets its own set i.e a collection of 8 disks across servers so what is happening is that you have objects are in huge variation i.e objects that landed on these sets 

```
~ grep data15 ~/Downloads/nodes.txt 
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   30G 100% /data15
/dev/sdo4             9.1T  9.1T   34G 100% /data15
```

In-fact inode usage is close to uniform across sets so that means that object spread is properly spread out, 
just that you have some objects that are too big i.e 

Baiscally one object that stays on the set 15 is 100x bigger than an object that stays on set 14 for example. 
Set determination happens based on object name to provide uniform distribution of objects but per object size 
is not spread across other sets. 

So what you are seeing is expected @Iguana9999phil - it happens that you have hit the worst case scenario
due to the type of objects you have and currently there is no way to address this problem without changing
the way we write to disks i.e changing our deterministic set based hashes for object names.

**Comment by harshavardhana** [comment]
I would suggest you split the 160G object into smaller constituent parts this would allow them to be more evenly spread out and the variation of maximum object size is not at the level of 100x than smallest object size. This would allow for size distribution to work properly. 

A fundamental design change is necessary for this to work without you spitting the objects, so that we make a link entry and have a fallback link elsewhere to take the I/O.  This would lead to double lookups but it is possible in theory.

**Comment by Iguana9999phil** [comment]
I've reviewed the docs before starting this project and I'm aware that objects stay in one set. What I don't understand is why the size of the object would matter / would make it use more space than set by EC:2. I can understand that some sets are using more space than others but shouldn't I still have the same usable space left?

**Comment by Iguana9999phil** [comment]
I know it's out of scope but is this what would roughly do what you suggested?
https://rclone.org/chunker/

**Comment by harshavardhana** [comment]
> I know it's out of scope but is this what would roughly do what you suggested?
> https://rclone.org/chunker/

No that won't help @Iguana9999phil - I am asking object names to be split as some parts and uploaded as different separate objects. 

> I've reviewed the docs before starting this project and I'm aware that objects stay in one set. What I don't understand is why the size of the object would matter / would make it use more space than set by EC:2. I can understand that some sets are using more space than others but shouldn't I still have the same usable space left?

IT would matter if it stays in one set since an object lets say is 1TiB in size goes to a set of 8 disks and another is 600mb goes to set of 8 disks the utilization would vary 100x. 

What is happening is that out of probability you have the object names that match the criteria where larger objects just happened to land on the same erasure set due to deterministic hash leading to disparity in usage. So now when you upload choosing the same set is not possible to write because that set has decreased in utilization due to a large differential in your object sizes.

**Comment by Iguana9999phil** [comment]
> > I know it's out of scope but is this what would roughly do what you suggested?
> > https://rclone.org/chunker/
> 
> No that won't help @Iguana9999phil - I am asking object names to be split as some parts and uploaded as different separate objects.
>

As stated (and correct me if i'm wrong) the issue is the big variation between object sizes, rclone with chunker would split those to a set (user adjustable) size so in theory wouldn't that solve the issue if I split everything to lets say 500MiB chunks? Minio will see those files of 500MiB chunks and can spread them within each set. 

I have no idea how to do splitting of object names etc. There was nothing in any of the documentation etc that suggest minio wouldn't be ideal for a dataset of media files. So what would be my options?

**Comment by harshavardhana** [comment]
> As stated (and correct me if i'm wrong) the issue is the big variation between object sizes, rclone with chunker would split those to a set (user adjustable) size so in theory wouldn't that solve the issue if I split everything to lets say 500MiB chunks? Minio will see those files of 500MiB chunks and can spread them within each set.

No, the hashing doesn't happen on individual parts by themselves as they belong to the same object @Iguana9999phil - all parts belonging to the same object stay on the same set. Otherwise, there will be too much overhead - too many calls for even simple data transfer. 

> I have no idea how to do splitting of object names etc. There was nothing in any of the documentation etc that suggest minio wouldn't be ideal for a dataset of media files. So what would be my options?

There is no option for now. This is a corner case that you seem to have hit - the way to address is by changing object names slightly by adding another heirarchy - You can figure this out using this python code. 

```py
#!/usr/bin/env python3

import argparse
import binascii
import siphash
import uuid
import os
import humanize

parser = argparse.ArgumentParser(description='Set indexer')
parser.add_argument('hash', metavar='HASH', type=str, help='type of hash used')
parser.add_argument('key', metavar='KEY', type=str, help='key name for the hash index')
parser.add_argument('setcount', metavar='SET', type=int, help='total set counts')

args = parser.parse_args()
if args.hash == 'crc32':
    print("Hash set index: '{0}'".format(binascii.crc32(args.key.encode()) % args.setcount))
else:
    sip = siphash.SipHash_2_4(uuid.UUID('{15e01db7-a3c0-4028-bc06-d8ca28c6de30}').bytes)
    sip.update(args.key.encode())
    print("Hash set index: '{0}' - size '{1}'".format(
        sip.hash() % args.setcount,
        humanize.naturalsize(os.stat(args.key).st_size))
    )
```

`your-deployment-id` can be obtained by using `jq .id backend-dirve/.minio.sys/format.json` 

and then use the python code 
```
~ find <your-data-set> -type f -exec python hash-set.py siphash {} 15 \;
```

This would give the distribution of sets and the size, at this point just adding one 
byte would change the entire sequence i.e as much as changing the character.

**Comment by Iguana9999phil** [comment]
Ok but I'm sure I'm not the first one who uses minio for media storage right? Is it an corner case due to amount of nodes / sets made? Trying to understand how a file, which gets spread over a set can balloon in size? 

As an example, looking at my second biggest directory which is on average 2.55TiB per disk so 306TiB raw in total for 180TiB real usage, that is a ratio of 1.7 so much higher than anticipated. Smaller directories are unaffected though.

**Comment by harshavardhana** [comment]
As I said I can't predict it at a whim @Iguana9999phil and it has really nothing much to do with number of sets or drives either. If you have more disks and more sets (erasure stripe sets) - it would change dramatically as well  - as there are more sets to hash to. 

It is just very very rare that is why you have seen it first - because of the high variance in object sizes as well - for the most part, average size of objects don't vary so much for us in practical deployments, but not that we don't know it when we designed it either. It is just very rare unfortunately it did happen to you so I gave a suggestion already on what to do. 

It may be unacceptable to you as a solution but that is what I have right now unless we fundamentally address it inside MinIO to allow for a rare corner case that can be triggered/exploited under certain conditions.

**Comment by harshavardhana** [comment]
> Ok but I'm sure I'm not the first one who uses minio for media storage right? Is it an corner case due to amount of nodes / sets made? Trying to understand how a file, which gets spread over a set can balloon in size?

It's not a ballon it's very simple 

- foo1 - 1TiB (goes to set 1)
- foo2 - 1Byte (goes to set 2)

By chance you happen to have another object `something-foo1` - that also happens to land on set 1 - even this object is 1TiB in size. 

i.e would mean that 

set1 now has `foo1`, `something-foo1` - 2TiB 

If your usable space across set1 was 2TiB - you ran out of disk space on set1, you do have plenty of space on set2 but now you come up with another object i.e `something-like-foo1` now this also, unfortunately, gets hashed to set1 - there is no space while you do have plenty of space on set2. 

There is a deterministic hash so that means such a set should be able to carry enough storage space for I/O - this doesn't happen in practice because hashes lead to uniform distribution across many many objects.  Like in your case each disk is approximately around `400k`  give or take here and there some transient objects created by list caching etc. 

But there is a rare adverse effect when some objects that happen to land on a particular set also had the requirement of large object sizes. This has never happened in the field while we always knew that this is potentially possible in some rare scenarios. 

Unfortunately, that is your scenario here - so a tiny change in object name would automatically land you to a newer set - but yes now storage exposes some gritty details to application which is not nice either.  Unless we handle it ourselves internally for such an eventuality it is not possible to easily by some tiny code change address this situation. 

That is why I gave you the python script to see for yourself where do those large objects land and how your object distribution looks like. It would give you an indication of what can be done to mitigate this problem interim. 

This has nothing to do with your storage class settings, erasure coding, or how you choose to deploy.

**Comment by harshavardhana** [comment]
> This has nothing to do with your storage class settings, erasure coding, or how you choose to deploy.

Although NOTE: changing 15 sets to 16 sets i.e one more drive in the deployment would change the hashing entirely and you may never see the problem.

**Comment by Iguana9999phil** [comment]
I haven't checked with the python script, but found the issue is only present in the 2 largest directories which also contain the most objects with the biggest file differences. 

I'll give up on trying to understand the reason for this issue as it's too deep for me to wrap my mind around, but thanks for trying to explain ;-) 

I'll do some more testing and think about how to solve it for now.
